src/structures/User.js

```diff
@@ -3,26 +3,21 @@ const Constants = require('../util/Constants');
 const { Presence } = require('./Presence');
 const UserProfile = require('./UserProfile');
 const Snowflake = require('../util/Snowflake');
+const Base = require('./Base');
 const { Error } = require('../errors');
 
 /**
  * Represents a user on Discord.
  * @implements {TextBasedChannel}
  */
-class User {
+class User extends Base {
   constructor(client, data) {
-    /**
-     * The client that created the instance of the user
-     * @name User#client
-     * @type {Client}
-     * @readonly
-     */
-    Object.defineProperty(this, 'client', { value: client });
-
-    if (data) this.setup(data);
+    super(client);
+    this._patch(data);
   }
 
-  setup(data) {
+  _patch(data) {
```

Comment:
PresenceUpdate packets (maybe others too) are calling this method with partial data, which somewhat breaks the user cache, for reference:
```js
User {
id: '173547401905176585',
username: undefined,
discriminator: undefined,
avatar: undefined,
bot: false,
lastMessageID: null,
lastMessage: null }
```
For example using the "old" `patch` method for patching with potential partial data and moving the initial property assigning to the constructor (or even to a/the setup method) would solve that issue.